### PR TITLE
E2e test for the event-logging

### PR DESCRIPTION
**Pull request by vlvasilev**
**How to categorize this PR?**

/area logging
/kind test

**What this PR does / why we need it**:
This PR adds an e2e test for the `event-logging` #6223 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other developer
E2e test for the `event-logging`
```

### CODE REVIEW

**Review by timebertt — CHANGES_REQUESTED** [review]
We generally don't want to invest into testmachinery tests anymore, see https://github.com/gardener/gardener/blob/master/docs/development/testing.md#test-machinery-tests, especially

> however, because of the listed drawbacks, we should rather focus on augmenting our e2e tests, as we can run them locally and in CI in order to catch bugs before they get merged
> it's still a good idea to add test machinery tests if a feature needs to be tested that is depending on some installation-specific configuration

It seems like there is no installation-specific config involved here and this is generally a good candidate for being tested in e2e (meaning on PR level and periodically on master).
Hence, I strongly suggest to rework this test and add it to the e2e test suite instead.

**Comment by vlvasilev** [comment]
> Hence, I strongly suggest to rework this test and add it to the e2e test suite instead.

So, I have to implement [PodExecByLabel](https://github.com/gardener/gardener/blob/195189a61500b1010d9555c816755a91c0747e7c/test/framework/k8s_utils.go#L229) function. If so where sould I put it? In the global [common.go](https://github.com/gardener/gardener/blob/master/test/e2e/common.go)?

**Comment by timebertt** [comment]
@vlvasilev Thanks for reworking the PR! Taking another look :)

File: test/e2e/shoot/logging/event_logging.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	var (
+		shootPurpose   = gardencorev1beta1.ShootPurposeDevelopment
+		shootNamespace string
+		lokiLabels     = map[string]string{
+			"app":  "loki",
+			"role": "logging",
+		}
+	)
+
+	f := defaultShootCreationFramework()
```
Comment (timebertt) [inline test/e2e/shoot/logging/event_logging.go]: Generally, I don't think it's necessary to add another test case and create another shoot just for verifying the `event-logger`.
Instead, I would rather add an additional verification step to an existing e2e test case, e.g. after https://github.com/gardener/gardener/blob/d340d78b17261862a5e1cc8b6648e9b58ab1314d/test/e2e/shoot/create_update_delete.go#L57-L63

You could move the verification logic into `internal/logging` and expose it as some kind of "verifier" similar to https://github.com/gardener/gardener/blob/d340d78b17261862a5e1cc8b6648e9b58ab1314d/test/e2e/shoot/internal/rotation/observability.go

File: test/e2e/shoot/logging/event_logging.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	})
+
+	It("should get events from loki", Label("event-logger"), func() {
+		ctx, cancel := context.WithTimeout(parentCtx, 15*time.Minute)
+		defer cancel()
+
+		By("Wait until Loki StatefulSet is ready")
+		Eventually(func(g Gomega) {
+			g.Expect(f.WaitUntilStatefulSetIsRunning(ctx, lokiName, shootNamespace, f.ShootFramework.SeedClient)).To(Succeed())
+		}).Should(Succeed())
```
Comment (timebertt) [inline test/e2e/shoot/logging/event_logging.go]: `WaitUntilStatefulSetIsRunning` already retries until the condition is met, so wrapping it in `Eventually` doesn't make much sense.
You could drop `Eventually` in this case.

File: test/e2e/shoot/logging/event_logging.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+		By("Create seed event")
+		seedEvent := getEventFor("seed", shootNamespace)
+		Expect(f.ShootFramework.SeedClient.Client().Create(ctx, seedEvent)).To(Succeed())
+
+		By("Create shoot event")
+		shootEvent := getEventFor("shoot", kubeSystemNamespace)
+		Expect(f.ShootFramework.ShootClient.Client().Create(ctx, shootEvent)).To(Succeed())
+
+		By("Wait until seed event is created")
+		framework.ExpectNoError(waitUntilEventIsCreated(ctx, f.Logger, seedEventName, shootNamespace, f.ShootFramework.SeedClient))
```
Comment (timebertt) [inline test/e2e/shoot/logging/event_logging.go]: Use `Expect(...).To(Succeed())` instead of `framework.ExpectNoError` to minimize confusion.

File: test/e2e/shoot/logging/event_logging.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+		event.Name = "unknown-event"
+		event.Message = "Event from Event Logger integration test related to unknown cluster type"
+	}
+
+	return event
+
+}
+
+// WaitUntilEventIsCreated waits until the given event is created
+func waitUntilEventIsCreated(ctx context.Context, logger logr.Logger, name string, namespace string, k8sClient kubernetes.Interface) error {
```
Comment (timebertt) [inline test/e2e/shoot/logging/event_logging.go]: I don't think there is any need to wait for the event to be created.

**Review by timebertt — CHANGES_REQUESTED** [review]

File: test/e2e/shoot/logging/common.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+// GetLokiLogsWithCMD gets logs from startTime until now for query to loki instance in <lokiNamespace>
+func GetLokiLogsWithCMD(ctx context.Context, logger logr.Logger, lokiLabels map[string]string, tenant, lokiNamespace, query string, startTime time.Time, client kubernetes.Interface) (*RangeSearchResponse, error) {
+	lokiLabelsSelector := labels.SelectorFromSet(labels.Set(lokiLabels))
+
+	if tenant == "" {
+		tenant = "fake"
+	}
+
+	command := fmt.Sprintf("wget 'http://localhost:%d/loki/api/v1/query_range?start=%d' -O- '--header=X-Scope-OrgID: %s' --post-data='query=%s'", lokiPort, startTime.UnixNano(), tenant, query)
```
Comment (timebertt) [inline test/e2e/shoot/logging/common.go]: Can we talk to loki directly via its `Ingress` instead of doing this hacky `wget` exec in the pod (which is not as stable)

Comment (vlvasilev) [reply test/e2e/shoot/logging/common.go]: Loki does not have ingress. The ingress which it has is only for pushing logs from the shoot node so in this case, Loki is not accessible outside the seed.

File: test/e2e/common.go [hunk 6]
```diff
@@ -39,6 +39,7 @@ func DefaultGardenConfig(projectNamespace string) *framework.GardenerConfig {
 
 // DefaultShoot returns a Shoot object with default values for the e2e tests.
 func DefaultShoot(name string) *gardencorev1beta1.Shoot {
+	purpose := gardencorev1beta1.ShootPurposeDevelopment
```
Comment (timebertt) [inline test/e2e/common.go]: Why do we need purpose development for the shoot?
Currently, the e2e shoots use purpose evaluation (default), but the event logging feature shouldn't depend on it.

File: test/e2e/shoot/create_update_delete.go [hunk 7]
```diff
@@ -62,6 +65,14 @@ var _ = Describe("Shoot Tests", Label("Shoot", "default"), func() {
 			g.Expect(shootClient.Client().List(ctx, &corev1.NamespaceList{})).To(Succeed())
 		}).Should(Succeed())
 
+		By("Verify the shoot event-logging")
+		Eventually(func(g Gomega) {
+			eventLoggerVerifier.Before(ctx)
+			eventLoggerVerifier.Prepare(ctx)
+			eventLoggerVerifier.Expect(ctx)
+			eventLoggerVerifier.After(ctx)
```
Comment (timebertt) [inline test/e2e/shoot/create_update_delete.go]: I don't think it makes sense to artificially press this verify into this `Before`, `Prepare`, ... structure. We have used it for the credentials rotation because it makes sense there.
But you can simplify this here and put everything into a single `Verify` method?

File: test/e2e/shoot/internal/logging/verifier.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+// limitations under the License.
+
+package logging
+
+import (
+	"context"
+)
+
+// Verifier does some assertions in different phases of the logging test.
+type Verifier interface {
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/verifier.go]: You don't need an interface for this. You can simply call the implementing type directly.

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+var (
+	lokiLabels = map[string]string{
+		"app":  "loki",
+		"role": "logging",
+	}
+)
+
+// EventLoggingVerifier verifies the event-logger.
+type EventLoggingVerifier struct {
+	*framework.ShootCreationFramework
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]:
```suggestion
	*framework.ShootFramework
```

`ShootFramework` should contain everything you need, right?

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+func (v *EventLoggingVerifier) After(ctx context.Context) {
+	By("Delete seed and shoot events")
+	Eventually(func(g Gomega) {
+		g.Expect(v.ShootFramework.SeedClient.Client().Delete(ctx, v.seedEvent)).To(Succeed())
+		g.Expect(v.ShootFramework.ShootClient.Client().Delete(ctx, v.shootEvent)).To(Succeed())
+	}).Should(Succeed())
+}
+
+func getEventFor(clusterType, namespace string) *corev1.Event {
+	clusterType = strings.ToLower(clusterType)
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]: Unneeded

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+
+	event := &corev1.Event{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+		},
+		InvolvedObject: corev1.ObjectReference{
+			Kind:      clusterType,
+			Namespace: namespace,
+			Name:      clusterType + "-" + eventLoggerName + "-testing",
+		},
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]: This doesn't make any sense. There is no such object in the system.
Either drop `InvolvedObject` entirely or reference an existing object.

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+		Reason: eventLoggerName + "-" + clusterType + "-testing",
+		Type:   "Normal",
+		Source: corev1.EventSource{
+			Component: eventLoggerName,
+			Host:      clusterType,
+		},
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]:
```suggestion
		Source: corev1.EventSource{
			Component: "event-logger-test"
		},
```

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+	eventLoggerName        = "event-logger"
+	kubeSystemNamespace    = "kube-system"
+	operatorID             = "operator"
+	lokiName               = "loki"
+	gardenNamesapce        = "garden"
+	fluentBitDaemonSetName = "fluent-bit"
+
+	defaultPollInterval = 5 * time.Second
+	lokiContainer       = "loki"
+	lokiPort            = 3100
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]: Why do you need so many constants? Can you double-check which ones can be inlined?

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+	v.seedEvent = getEventFor("seed", v.shootNamespace)
+	v.shootEvent = getEventFor("shoot", kubeSystemNamespace)
+
+	Eventually(func(g Gomega) {
+		g.Expect(v.ShootFramework.SeedClient.Client().Get(ctx, client.ObjectKey{Namespace: gardenNamesapce, Name: fluentBitDaemonSetName}, &appsv1.DaemonSet{})).To(Succeed())
+		g.Expect(v.ShootFramework.SeedClient.Client().Get(ctx, client.ObjectKey{Namespace: v.shootNamespace, Name: lokiName}, &appsv1.StatefulSet{})).To(Succeed())
+		g.Expect(v.ShootFramework.SeedClient.Client().Get(ctx, client.ObjectKey{Namespace: v.shootNamespace, Name: eventLogger}, &appsv1.Deployment{})).To(Succeed())
+	}).Should(Succeed())
+
+	By("Waite for the shoot Loki and Event-Logger to become ready")
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]: Typo
```suggestion
	By("Wait for the shoot Loki and Event-Logger to become ready")
```

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+	By("Wait until Loki receive shoot event")
+	Expect(waitUntilLokiReceivesEvent(ctx, v.Logger, 30*time.Second, lokiLabels, operatorID, v.shootNamespace, `origin_extracted="shoot",source="event-logger",sourceHost="shoot"`, []string{eventMessageForShoot}, time.Now().Add(-5*time.Minute), v.ShootFramework.SeedClient)).To(Succeed())
+}
+
+// After cleans all resources created in Prepare function
+func (v *EventLoggingVerifier) After(ctx context.Context) {
+	By("Delete seed and shoot events")
+	Eventually(func(g Gomega) {
+		g.Expect(v.ShootFramework.SeedClient.Client().Delete(ctx, v.seedEvent)).To(Succeed())
+		g.Expect(v.ShootFramework.ShootClient.Client().Delete(ctx, v.shootEvent)).To(Succeed())
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]:
```suggestion
		g.Expect(v.ShootFramework.SeedClient.Client().Delete(ctx, v.seedEvent)).To(Or(Succeed(), BeNotFoundError()))
		g.Expect(v.ShootFramework.ShootClient.Client().Delete(ctx, v.shootEvent)).To(Or(Succeed(), BeNotFoundError()))
```

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+	By("Wait until Loki receive seed event")
+	Expect(waitUntilLokiReceivesEvent(ctx, v.Logger, 30*time.Second, lokiLabels, operatorID, v.shootNamespace, `origin_extracted="seed",source="event-logger",sourceHost="seed"`, []string{eventMessageForSeed}, time.Now().Add(-5*time.Minute), v.ShootFramework.SeedClient)).To(Succeed())
+
+	By("Wait until Loki receive shoot event")
+	Expect(waitUntilLokiReceivesEvent(ctx, v.Logger, 30*time.Second, lokiLabels, operatorID, v.shootNamespace, `origin_extracted="shoot",source="event-logger",sourceHost="shoot"`, []string{eventMessageForShoot}, time.Now().Add(-5*time.Minute), v.ShootFramework.SeedClient)).To(Succeed())
+}
+
+// After cleans all resources created in Prepare function
+func (v *EventLoggingVerifier) After(ctx context.Context) {
+	By("Delete seed and shoot events")
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]: You can wrap this cleanup in `DeferCleanup`

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+		}
+
+		if len(search.Data.Result) < len(wantedEventMessages) {
+			log.Info("Waiting to receive all expected logs", "events", len(search.Data.Result), "wanted", len(wantedEventMessages))
+			return retry.MinorError(fmt.Errorf("received only %d/%d events", len(search.Data.Result), len(wantedEventMessages)))
+		} else if len(search.Data.Result) > len(wantedEventMessages) {
+			return retry.SevereError(fmt.Errorf("expected to receive %d events but was %d", len(wantedEventMessages), len(search.Data.Result)))
+		}
+
+		var aggErr errorsutil.AggregateError
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go:213]: I don't think this is the type you wanted to import, it's coming from `onsi/gomega`

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+	}).Should(Succeed())
+}
+
+// Expect is the the process were we expect to get the correct result from the logging test.
+func (v *EventLoggingVerifier) Expect(ctx context.Context) {
+	By("Wait until Loki receive seed event")
+	Expect(waitUntilLokiReceivesEvent(ctx, v.Logger, 30*time.Second, lokiLabels, operatorID, v.shootNamespace, `origin_extracted="seed",source="event-logger",sourceHost="seed"`, []string{eventMessageForSeed}, time.Now().Add(-5*time.Minute), v.ShootFramework.SeedClient)).To(Succeed())
+
+	By("Wait until Loki receive shoot event")
+	Expect(waitUntilLokiReceivesEvent(ctx, v.Logger, 30*time.Second, lokiLabels, operatorID, v.shootNamespace, `origin_extracted="shoot",source="event-logger",sourceHost="shoot"`, []string{eventMessageForShoot}, time.Now().Add(-5*time.Minute), v.ShootFramework.SeedClient)).To(Succeed())
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]: There are many parameters that always have the same value. Please drop those parameters.

Also, instead of passing `time.Now().Add(-5*time.Minute)`, should we pass the events' `creationTimestamp`?

**Review by timebertt — CHANGES_REQUESTED** [review]

File: test/e2e/common.go (hunk 6 above)
Comment (vlvasilev) [reply test/e2e/common.go]: I didn't know that the default is 'evaluation'. Thanks.

File: test/e2e/shoot/create_update_delete.go (hunk 7 above)
Comment (vlvasilev) [reply test/e2e/shoot/create_update_delete.go]: As you wish. I just followed your advice to make a vast interface as the in the example which you gave me.

File: test/e2e/shoot/internal/logging/verifier.go (hunk 8 above)
Comment (vlvasilev) [reply test/e2e/shoot/internal/logging/verifier.go]: but what if there are multiple logging e2e tests?

File: test/e2e/shoot/internal/logging/event_logging.go (hunk 11 above)
Comment (vlvasilev) [reply test/e2e/shoot/internal/logging/event_logging.go]: I see that the InvolvedObject is not optional so I guess I cannot omit it.
And I am not searching for logical sense, I need this resource only for testing purposes. Could you tell me which object in the shoot should I address when creating the event?

**Comment by vlvasilev** [comment]
@timebertt could you please help me with that DeferCleanup() function?
I can't understand why is failing. Do I use it wrongly?

**Comment by timebertt** [comment]
@vlvasilev please add new commits to the PR when addressing review comments. It's super hard to follow changes when they are force-pushed.
Prow will squash the commits for you when merging the PR automatically.

File: test/e2e/shoot/create_update_delete.go [hunk 19]
```diff
@@ -62,6 +65,12 @@ var _ = Describe("Shoot Tests", Label("Shoot", "default"), func() {
 			g.Expect(shootClient.Client().List(ctx, &corev1.NamespaceList{})).To(Succeed())
 		}).Should(Succeed())
 
+		By("Verify the shoot event-logging")
+		DeferCleanup(func() {
+			eventLoggerVerifier.After(ctx)
```
Comment (timebertt) [inline test/e2e/shoot/create_update_delete.go]: `ctx` will be canceled once the cleanup action is executed.
You could derive a new context from `parentContext` for cleanup.

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+
+// Verify verifies that the event logging is working properly
+func (v *EventLoggingVerifier) Verify(ctx context.Context) {
+	v.before(ctx)
+	v.prepare(ctx)
+	v.expect(ctx)
+}
+
+// After cleans all resources created in prepare function
+func (v *EventLoggingVerifier) After(ctx context.Context) {
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]: Maybe call this `Cleanup`?

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+func (v *EventLoggingVerifier) prepare(ctx context.Context) {
+	By("Create seed and shoot events")
+	Eventually(func(g Gomega) {
+		g.Expect(v.ShootFramework.SeedClient.Client().Create(ctx, v.seedEvent)).To(Succeed())
+		g.Expect(v.ShootFramework.ShootClient.Client().Create(ctx, v.shootEvent)).To(Succeed())
+	}).Should(Succeed())
+}
+
+// expect is the the process were we expect to get the correct result from the logging test.
+func (v *EventLoggingVerifier) expect(ctx context.Context) {
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go:135]: I think this test is not precise enough. It should expect that the exact event is collected by the logging stack that was created earlier on.
Otherwise, the test might succeed falsely when executed multiple times in a row.

File: test/e2e/shoot/internal/logging/event_logging.go (hunk 17 above)
Comment (timebertt) [reply test/e2e/shoot/internal/logging/event_logging.go:213]: This is not resolved.

File: test/e2e/shoot/internal/logging/event_logging.go [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+	v.prepare(ctx)
+	v.expect(ctx)
+}
+
+// After cleans all resources created in prepare function
+func (v *EventLoggingVerifier) After(ctx context.Context) {
+	By("Delete seed and shoot events")
+	Eventually(func(g Gomega) {
+		g.Expect(v.ShootFramework.SeedClient.Client().Delete(ctx, v.seedEvent)).To(Or(Succeed(), BeNotFoundError()))
+		g.Expect(v.ShootFramework.ShootClient.Client().Delete(ctx, v.shootEvent)).To(Or(Succeed(), BeNotFoundError()))
```
Comment (timebertt) [inline test/e2e/shoot/internal/logging/event_logging.go]: The shoot might already be deleted when this is executed.
Please make sure it will succeed in this case as well.

**Review by timebertt — COMMENTED** [review]
This looks better now, a few remaining comments.

**Comment by timebertt** [comment]
/unassign
as I will be unavailable over the next weeks

**Comment by rfranzke** [comment]
/assign
@vlvasilev Can you please update the PR?

**Comment by vlvasilev** [comment]
>  Can you please update the PR?

Sorry, I was on a vacation and I will be next week, too.
I will try to do it in the last two days of that week but I am not familiar with the e2e tests and I may not succeed.

File: test/e2e/shoot/internal/logging/event_logging.go (hunk 22 above)
Comment (vlvasilev) [reply test/e2e/shoot/internal/logging/event_logging.go]: How?
